client/my-sites/domains/domain-management/list/bulk-edit-contact-info.jsx

```diff
@@ -0,0 +1,218 @@
+
+/**
+ * Internal dependencies
+ */
+import DomainContactDetails from 'calypso/my-sites/checkout/composite-checkout/components/domain-contact-details';
+import { Button } from '@automattic/components';
+import { fetchUserSettings } from 'calypso/state/user-settings/actions';
+import { recordTracksEvent } from 'calypso/state/analytics/actions';
+import getContactDetailsCache from 'calypso/state/selectors/get-contact-details-cache';
+import isRequestingContactDetailsCache from 'calypso/state/selectors/is-requesting-contact-details-cache';
+import { requestContactDetailsCache, saveWhois } from 'calypso/state/domains/management/actions';
+import getUserSettings from 'calypso/state/selectors/get-user-settings';
+import { errorNotice, infoNotice, successNotice } from 'calypso/state/notices/actions';
+import { UPDATE_CONTACT_INFORMATION_EMAIL_OR_NAME_CHANGES } from 'calypso/lib/url/support';
+import FormLabel from 'calypso/components/forms/form-label';
+import FormCheckbox from 'calypso/components/forms/form-checkbox';
+import DesignatedAgentNotice from 'calypso/my-sites/domains/domain-management/components/designated-agent-notice';
+import { isFetchingUserSettings } from 'calypso/state/user-settings/selectors';
+
+import wp from 'calypso/lib/wp';
+
+const wpcom = wp.undocumented();
+
+/**
+ * Style dependencies
+ */
+import './list-all.scss';
+
+class BulkEditContactInfo extends React.Component {
+	static propTypes = {
+		handleSaveContactInfo: PropTypes.func.isRequired,
+		isDisabled: PropTypes.bool,
+		isSubmitting: PropTypes.bool,
+		onTransferLockOptOutChange: PropTypes.func.isRequired,
+		emailOnly: PropTypes.bool,
+	};
+
+	static defaultProps = {
+		isDisabled: false,
+		isSubmitting: false,
+		emailOnly: false,
+	};
+
+	state = {
+		hasSetContactDetailsFromCache: false,
+		contactDetails: {},
+		errorMessages: {},
+	};
+
+	componentDidMount() {
+		if ( isEmpty( this.props.userSettings ) ) {
+			this.props.fetchUserSettings();
+		}
+
+		if ( isEmpty( this.props.contactDetailsCache ) ) {
+			this.props.requestContactDetailsCache();
+		}
+
+		this.debouncedValidateContactDetails = debounce( this.validateContactDetails, 500 );
+	}
+
+	componentDidUpdate() {
+		const email = ! isEmpty( this.props.userSettings?.new_user_email )
+			? this.props.userSettings?.new_user_email
+			: this.props.userSettings?.user_email;
+
+		if ( ! isEmpty( email ) && isEmpty( this.state.contactDetails?.email ) ) {
+			this.updateDomainContactFields( { email: email } );
+		}
+
+		if (
+			! isEmpty( this.props.contactDetailsCache ) &&
+			false === this.state.hasSetContactDetailsFromCache
+		) {
+			this.setContactDetailsFromCache( this.props.contactDetailsCache );
+		}
+	}
+
+	isLoading() {
+		return (
+			isEmpty( this.props.contactDetailsCache ) ||
+			this.props.isRequestingContactDetailsCache ||
+			this.props.isFetchingUserSettings
+		);
+	}
+
+	setContactDetailsFromCache = ( data ) => {
+		delete data.email;
```

Comment:
That's a pretty big side effect considering that this data comes from this component's props, so this will change the object on the parent component. Is there any non-mutation way to do this?

Reply:
Oof. Another good point.  I'll make a copy of this data to use inside this component.